### PR TITLE
tikv: make pushed txn judge as pushed in the same lock-resolve batch

### LINKED ISSUE #21658: Queries using batch point get plan may be blocked when encounters locks of another alive transaction
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Run `CaseRCResolveLockBatchPointGet` which is implemented [here](https://github.com/pingcap/automated-tests/pull/543/files#diff-5235c0b2974a3c43dfcd805890740fb1a151a182d3162666bba3ee748893d211R1631), and with TiDB's failpoint API enabled.

The procedure of the case:

```sql
create table t (id int primary key, v int);
insert into t values (0, 0), (1, 10), (2, 20), (3, 30);
```

| session1 | session2 | result/comment |
|---|---|---|
| `set transaction isolation level read commited; begin pessimistic;` | | |
| `select id, v from t where id = 1 or id = 2 or id = 3 order by id;` | | `(1, 10), (2, 20), (3, 30)` |
| | `begin pessimistic;` | |
| | `update t set v = v + 1 where id = 0;` | |
| | `update t set v = v + 1 where id > 0;` | |
| | | Enable failpoint `github.com/pingcap/tidb/store/tikv/beforeCommit` as `"pause"` at this time |
| | `commit` | It will be blocked due to the failpoint. Don't wait for it and continue following steps. |
| `select id, v from t where id = 1 or id = 2 or id = 3 order by id;` | | Blocked and maybe report timeout error. |

### 2. What did you expect to see? (Required)

In the last step above, session 1's transaction should push session2's transaction's `min_commit_ts`, and continue reading the old data bypassing the locks.

### 3. What did you see instead (Required)

Session 1 blocked and reports timeout error:

```
[2020/12/09 11:08:23.827 +00:00] [WARN] [backoff.go:342] ["txnLockFast backoffer.maxSleep 40000ms is exceeded, errors:\nbatchGet lockedKeys: 3 at 2020-12-09T11:08:19.652431113Z\nbatchGet lockedKeys: 3 at 2020-12-09T11:08:21.209817741Z\nbatchGet lockedKeys: 3 at 2020-12-09T11:08:23.827054497Z"]
[2020/12/09 11:08:23.827 +00:00] [INFO] [conn.go:809] ["command dispatched failed"] [conn=577362351838199857] [connInfo="id:577362351838199857, addr:10.9.85.17:58554 status:11, collation:utf8mb4_general_ci, user:root"] [command=Query] [status="inTxn:1, autocommit:1"] [sql="select id, v from t where id = 1 or id = 2 or id = 3 order by id"] [txn_mode=PESSIMISTIC] [err="[tikv:9004]Resolve lock timeout\ngithub.com/pingcap/errors.AddStack\n\t/home/zyguan/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20201126102027-b0a155152ca3/errors.go:174\ngithub.com/pingcap/errors.Trace\n\t/home/zyguan/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20201126102027-b0a155152ca3/juju_adaptor.go:15\ngithub.com/pingcap/tidb/store/tikv.(*tikvSnapshot).batchGetSingleRegion\n\t/home/zyguan/go/src/github.com/pingcap/tidb/store/tikv/snapshot.go:346\ngithub.com/pingcap/tidb/store/tikv.(*tikvSnapshot).batchGetKeysByRegions\n\t/home/zyguan/go/src/github.com/pingcap/tidb/store/tikv/snapshot.go:243\ngithub.com/pingcap/tidb/store/tikv.(*tikvSnapshot).BatchGet\n\t/home/zyguan/go/src/github.com/pingcap/tidb/store/tikv/snapshot.go:153\ngithub.com/pingcap/tidb/kv.(*BufferBatchGetter).BatchGet\n\t/home/zyguan/go/src/github.com/pingcap/tidb/kv/utils.go:102\ngithub.com/pingcap/tidb/executor.(*BatchPointGetExec).initialize\n\t/home/zyguan/go/src/github.com/pingcap/tidb/executor/batch_point_get.go:340\ngithub.com/pingcap/tidb/executor.(*BatchPointGetExec).Next\n\t/home/zyguan/go/src/github.com/pingcap/tidb/executor/batch_point_get.go:143\ngithub.com/pingcap/tidb/executor.Next\n\t/home/zyguan/go/src/github.com/pingcap/tidb/executor/executor.go:277\ngithub.com/pingcap/tidb/executor.(*recordSet).Next\n\t/home/zyguan/go/src/github.com/pingcap/tidb/executor/adapter.go:129\ngithub.com/pingcap/tidb/server.(*tidbResultSet).Next\n\t/home/zyguan/go/src/github.com/pingcap/tidb/server/driver_tidb.go:298\ngithub.com/pingcap/tidb/server.(*clientConn).writeChunks\n\t/home/zyguan/go/src/github.com/pingcap/tidb/server/conn.go:1728\ngithub.com/pingcap/tidb/server.(*clientConn).writeResultset\n\t/home/zyguan/go/src/github.com/pingcap/tidb/server/conn.go:1688\ngithub.com/pingcap/tidb/server.(*clientConn).handleStmt\n\t/home/zyguan/go/src/github.com/pingcap/tidb/server/conn.go:1585\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuery\n\t/home/zyguan/go/src/github.com/pingcap/tidb/server/conn.go:1451\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch\n\t/home/zyguan/go/src/github.com/pingcap/tidb/server/conn.go:1027\ngithub.com/pingcap/tidb/server.(*clientConn).Run\n\t/home/zyguan/go/src/github.com/pingcap/tidb/server/conn.go:792\ngithub.com/pingcap/tidb/server.(*Server).onConn\n\t/home/zyguan/go/src/github.com/pingcap/tidb/server/server.go:459\nruntime.goexit\n\t/usr/local/lib/go/src/runtime/asm_amd64.s:1373"]
[2020/12/09 11:08:23.880 +00:00] [INFO] [commit.go:92] ["2PC commitTS rejected by TiKV, retry with a newer commitTS"] [conn=577362351838199859] [txnStartTS=421399641765707777] [info="{StartTs:421399641765707777 AttemptedCommitTs:421399641831243777 Key:7480000000000000465f728000000000000000 MinCommitTs:421399642355531778}"]
```

### 4. What is your TiDB version? (Required)

master with some not-related changes (ba9789eed17cc9351b71b3507a3f0e4c23efc28b from https://github.com/pingcap/tidb/pull/21557), built with failpoints enabled.

### PR DESCRIPTION
**Pull request by lysu**
Signed-off-by: lysu <sulifx@gmail.com>

### What problem does this PR solve?

Issue Number: close #21658

Problem Summary:

https://github.com/tikv/tikv/pull/8751 change the `TxnStatus.status` return value in tikv.

for example: a txn1 write 3 locks in lock1, lock2, lock3 and a batch query meet those 3 locks and try to resolve.

- in previous: both lock1-lock3's resolve request will return `TxnStatus.status` with `Action_MinCommitTSPushed`
- but now: only lock1 will return `Action_MinCommitTSPushed`, lock2, lock3 will return `Action_NoAction` and this make tidb judge as cannot pushed.

current, we use [clientHelper](https://github.com/pingcap/tidb/blob/1b12071e524bad1660dbacfa0f18cd9cd6b602d6/store/tikv/coprocessor.go#L948) to avoid pushed lock be meeting in following query, but miss push the pushed lock in the same resolve lock batch, I'm not sure whether tikv can identify pushed-locks from locks easily...but it seems tidb have information about it and can handle it too.

> ~(maybe more aggressive, we can avoid send duplicate resolve lock request if lock's txn has be pushed???  - - ).~

### What is changed and how it works?

What's Changed:

for same txn, `Action_NoAction` after `Action_MinCommitTSPushed` shouldn't lead to judge as push fail..

How it Works:

check is it be pushed in previous

### Related changes

- n/a

### Check List 

Tests 

- Integration test: will failed with tikv in [check_dev_2](https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tidb_ghpr_check_2/detail/tidb_ghpr_check_2/68174/pipeline/53#step-75-log-442) without this PR using master tikv

Side effects

- n/a

### Release note 

-  Fix push commit-ts failure when met more than one lock belong to the same transaction in the same resolve-lock batch

### CODE REVIEW

**Comment by MyonKeminta** [comment]
Oh sorry, I didn't notice you already done this and still trying to investigate it 😂

**Comment by cfzjywxk** [comment]
LGTM

File: store/tikv/lock_resolver.go [hunk 1]
```diff
@@ -421,10 +421,12 @@ func (lr *LockResolver) resolveLocks(bo *Backoffer, callerStartTS uint64, locks
 				}
 			} else {
 				if status.action != kvrpcpb.Action_MinCommitTSPushed {
-					pushFail = true
-					return nil
+					if _, exists := pushed[l.TxnID]; !exists {
```
Comment (youjiali1995) [inline store/tikv/lock_resolver.go:424]: It's a bug in TiKV. Let me fix it. However, we can avoid pushing locks of the same transaction with the idea here.

Comment (cfzjywxk) [reply store/tikv/lock_resolver.go:424]: Do we have an issue for it?

Comment (youjiali1995) [reply store/tikv/lock_resolver.go:424]: No issue in TiKV repo. https://github.com/pingcap/tidb/issues/21658 is enough and this bug only exists in v5.0.

Comment (youjiali1995) [reply store/tikv/lock_resolver.go:424]: https://github.com/tikv/tikv/pull/9319

Comment (sticnarf) [reply store/tikv/lock_resolver.go:424]: Is it change necessary if tikv/tikv#9319 is merged?

Comment (lysu) [reply store/tikv/lock_resolver.go:424]: I think it seems no necessary - - but maybe we can try and test avoid pushing locks in future

**Pull request closed without merging**